<launch>
  <arg name="ns" default="v1"/>
  <arg name="partner" default="v2"/>
  <arg name="front_back_baseline" default="1"/>
  <arg name="initialPoseX" default="1.0"/>
  <arg name="initialPoseY" default="1.0"/>
  <arg name="initialPoseTheta" default="0.0"/>
  <arg name="agents" default="2"/>
  <arg name="id" default="1"/>
  <arg name="use_simulator" default="true" />
  <arg name="use_localization" default="true"/>

  <arg name="port" default="/dev/actuators/tfrog" />		
  <arg name="top_port_lrf" default="192.168.150.10" />
  <arg name="tough_port_lrf" default="192.168.0.10" />
  <arg name="port_ip" default="10940" />

  <arg name="mapdir" default="/home/$(env USER)/maps/2017_10_26_around_nic/"/>

  <group ns="$(arg ns)">
    <!--Launch agent-->
    <node pkg="sbs_motion" type="sbs_agent" name ="agent" output="screen" >
      <param name="partner" value="$(arg partner)"/>
      <param name="agents" value="$(arg agents)"/>
      <param name="id" value="$(arg id)"/>		
      <param name="initialPoseX" value="$(arg initialPoseX)"/>
      <param name="initialPoseY" value="$(arg initialPoseY)"/>
      <rosparam command="load" file="$(find sbs_motion)/launch/util.yaml" />
    </node>

    <remap from="/vehicle_acceleration" to="$(arg ns)/vehicle_acceleration"/>
    <remap from="/vehicle_yawrate" to="$(arg ns)/vehicle_yawrate"/>
    <remap from="/vehicle_speed" to="$(arg ns)/vehicle_speed"/>
    <remap from="/current_pose" to="$(arg ns)/current_pose"/>
    <remap from="/filtered_ndt_current_pose" to="$(arg ns)/filtered_ndt_current_pose"/>
    

    <!--<node pkg="nodelet" type="nodelet" name="base_nodelet_manager" args="manager" output="log" />
    <node pkg="nodelet" type="nodelet" name="cmd_vel_mux" args="load yocs_cmd_vel_mux/CmdVelMuxNodelet base_nodelet_manager" output="log" >
      <param name="yaml_cfg_file" value="$(find ypspur_ros_bridge)/config/cmd_mux_config.yaml" />
      <remap from="cmd_vel_mux/output" to="cmd_vel" />
      <remap from="output" to="cmd_vel" />
     
      
    </node>-->
    <group unless="$(arg use_simulator)">
       <include file="$(find ndt3d)/launch/localization.launch">
	 <arg name="mapdir" value="$(arg mapdir)"/>
	 <arg name="tf_prefix" value="$(arg ns)" />
       </include>

      <!-- TOP LRF -->
      <node name="top_urg_node" pkg="urg_node" type="urg_node" unless="$(arg use_log)">
	<param name="ip_address" value="$(arg top_port_lrf)" type="string" />
	<param name="ip_port" value="$(arg port_ip)" type="string" />
	<param name="frame_id" type="string" value="laser" />
	<param name="angle_min" value="-2.0" />
	<param name="angle_max" value="1.0" />
	<param name="calibrate_time" type="bool" value="true" />
	<remap from="/scan" to="/scan_orig0" />
	<param name="publish_multiecho" type="bool" value="false" />
      </node>
      
      <!-- DRIVER -->
      <node pkg="rrc_controller" type="rrc_controller" name="rrc_controller" respawn="true">
	<param name="port"   value="$(arg port)" type="string" /> 
	<param name="vel"    value="1.0" type="double" /> 
	<param name="angvel" value="1.0" type="double" /> 
	<param name="acc"    value="0.7" type="double" /> 
	<param name="angacc" value="1.5" type="double" /> 
	<param name="hz" value="50.0" type="double" /> 
	<param name="origin_id" value="" type="string" />
	<remap from="/driver/odom" to="odom" />
	<remap from="/ypspur_driver/cmd_vel" to="cmd_vel" />
      </node>
      
      <!-- LOCALIZATION -->
      <group if="$(arg use_localization)">
	<group if="$(arg use_initial_pose)">
	  <rosparam param="initial_pose_x" ns="amcl">0</rosparam>
	  <rosparam param="initial_pose_y" ns="amcl">0</rosparam>
	  <rosparam param="initial_pose_a" ns="amcl">0</rosparam>
	</group>
	<node name="amcl" pkg="amcl" type="amcl">
	  <param name="odom_alpha1" type="double" value="2" />
	  <param name="odom_alpha2" type="double" value="2" />
	  <param name="odom_alpha3" type="double" value="2" />
	  <param name="odom_alpha4" type="double" value="2" />
	  <param name="laser_max_beams" type="int" value="200" />
	  <param name="update_min_d" type="double" value="0.1" />
	  <param name="update_min_a" type="double" value="0.2" />
	  <param name="use_map_topic" type="bool" value="true" />
	  <param name="laser_model_type" type="string" value="beam" />
	  <param name="min_particles" type="int" value="150" />
	  <param name="max_particles" type="int" value="300" />
	  <param name="laser_likelihood_max_dist" type="double" value="1.0" />
	  <param name="odom_frame_id" type="string" value="$(arg ns)/odom" />
	  <param name="base_frame_id" type="string" value="$(arg ns)/base_link" />

	</node>
      </group>
    </group>
    <!-- LAUNCH SIMULATOR -->
    <group if="$(arg use_simulator)">
      <include file="$(find simple_simulator)/launch/simulator_wc.launch" >
	<arg name="tf_prefix" value="$(arg ns)" />
	<arg name="use_localization" value="$(arg use_localization)" />
	<arg name="delta_x" value="$(arg initialPoseX)" />
	<arg name="delta_y" value="$(arg initialPoseY)" />
	<arg name="delta_yaw" value="$(arg initialPoseTheta)" />
      </include>    
    </group>    
  </group>
</launch>
